### ISSUE TITLE
Encryption support

### ISSUE THREAD

**Issue by pwmorreale** [open-policy-agent/opa#1131]
A useful enhancement to OPA would be builtin functions enabling support for JOSE.

The use case is simply providing a means for a non-secure application to provide secret input to OPA.  As a configuration data format, JSON dovetails quite nicely with OPA and rego, enabling support for encrypted tokens provides a system with some guarantees regarding non-repudiation, confidentiality, and integrity.

From an API perspective, something similar to the current _io,jwt.decode.verify()_ builtin might work well.  From a user perspective, access to the jwt components after successful decrypt/verification should be sufficient.   Since this is a security issue, direct control (within the api?) on whether detected errors are written to logs would be desirable. 

Suggested supported algorithms include:
- A256KW
- A256GCMKW
- RSA-OAEP-256
would be a good start.

**Comment by tsandall** [comment]
@pwmorreale thanks for opening this. Just for my understanding, the non-secure application will supply the encrypted token to OPA in the input and the policy should be able to verify, decrypt, and decode the token. What's currently missing from OPA is:

- Ability to decrypt
- Guarantee that sensitive data is not being leaked (via logs, error responses, etc.)

**Comment by pwmorreale** [comment]
Precisely.  The jwt enters into OPA via an input stream.  The non-secure application is merely the transport for passing the secure datum to OPA.  

Logging presents its own set of issues...  ;-)   Given that the jwt in question is secure, logging it is not a security issue.  That said, jwt's can be several K's of bytes long and both hammer performance as well as fill up persistent resources.  (most especially under a DOS attack)

I think a longer term solution will be a more flexible logging implementation that allows the admin to configure which fields are included in a log msg, as well as (perhaps) a max byte count per log msg.

But that is another issue for another day... :-)

**Comment by repenno** [comment]
Looked at this and have a q: where would the necessary private/symmetric keys to decrypt the contents (for each JWT) come from?

**Comment by repenno** [comment]
Please allow me to expand. The cert to verify comes as part of the built in functions.  I believe the algorithms you suggested are based pub/pvt keys (https://tools.ietf.org/html/rfc7516, https://tools.ietf.org/html/rfc7520), but in general there is the issue of symmetric keys. Should they appear in the configuration? Come from Env variables? What would be your suggestion?

**Comment by pwmorreale** [comment]
The keys could come from either.   In the configuration scenario, one has the option to add authz security to the configuration to control access to the data document.

**Comment by deepahoskote** [comment]
@repenno , @tsandall   -  We are in need of this exact same feature request. 
Has this been implemented currently?

**Comment by deepahoskote** [comment]
@pwmorreale - Did you address this issue on your end with a workaround (extending OPA with a custom function, maybe)? Appreciate anything you might have to share.

**Comment by tsandall** [comment]
@deepahoskote no, as far as I'm aware, these built-in functions were never added to upstream OPA. That said, if this is an area that you'd like to contribute to, let us know.

cc @srenatus @anderseknert - we're using an internal fork (./internal/jwx) of https://github.com/lestrrat/go-jwx which is now archived and moved to https://github.com/lestrrat-go/jwx. It doesn't look like we have the encryption packages forked so we'd have to update the internal fork to include them.

**Comment by anderseknert** [comment]
Sounds good to me. We'll need to be careful with the upgrade since we've made a few changes to that code over the years (PKCS8 encoded private keys, BuiltinContext random source, decode RSA private key into JWK format, and so on). Hopefully all covered by unit tests (😅) but could be worth an extra review to make sure we don't introduce regressions. Could also be that some of the things we've added are now covered by the library itself in the current version.

**Comment by sspaink** [comment]
The internal fork has been migrated to upstream (since v1.10): https://github.com/open-policy-agent/opa/pull/7733, which has the requested algorithms.

It has been a while, @deepahoskote or @pwmorreale do you still need this feature in OPA?

**Comment by pwmorreale** [comment]
I do not. Thank you.  Best Regards,-PWMOn Sep 8, 2026, at 10:56 AM, Sebastian Spaink ***@***.***> wrote:﻿sspaink left a comment (open-policy-agent/opa#1131)
The internal fork has been migrated to upstream (since v1.10): #7733, which has the requested algorithms.
It has been a while, @deepahoskote or @pwmorreale do you still need this feature in OPA?
